### PR TITLE
Sliding window local attention

### PR DESCRIPTION
**Pull request by changlan**
* Added sliding window local attention feature to MultiheadAttention and FlashAttention.
* Added or_masks and and_masks which would be useful to composite different mask functions.
* Fixed a bug that causes TPU decoding always to fail.

### CODE REVIEW

File: axlearn/common/attention.py [hunk 1]
```diff
@@ -1781,8 +1832,6 @@ def _forward_for_mode(
                 "key and value must be both None or both set, "
                 f"key:{type(key)}, value:{type(value)}"
             )
-        if self._mask_fn and (key is not None or value is not None):
-            raise ValueError("key and value are not expected when using an attention mask.")
```
Comment (ruomingp) [inline axlearn/common/attention.py:1785]: Can you remind me why this logic is removed?

**Review by ruomingp — APPROVED** [review]

File: axlearn/common/attention.py (hunk 1 above)
Comment (changlan) [reply axlearn/common/attention.py:1785]: It breaks [this test](https://github.com/apple/axlearn/blob/9ba319391c9019df2ff56e192320c45c8808c166/axlearn/common/flash_attention/layer_test.py#L498) where `key` and `value` are specified. It doesn't seem to me that this restriction is necessary, as `MaskFn` does expect `key_position` as well. WDYT?

**Review by markblee — APPROVED** [review]

File: axlearn/common/attention.py (hunk 1 above)
Comment (ruomingp) [reply axlearn/common/attention.py:1785]: Got it. Thanks!

File: axlearn/common/flash_attention/tpu_attention_test.py [hunk 2]
```diff
@@ -49,6 +50,19 @@ class TestFlashAttention(TestCase):
         ),
     ]
 
+    @parameterized.product(seq_len=[8, 16, 32, 128], sliding_window_size=[4, 8, 16])
+    def test_sliding_window_mask(self, seq_len, sliding_window_size):
+        shape = (seq_len, seq_len)
+        ref_mask = splash_attention_mask.LocalMask(
+            shape=shape, window_size=(sliding_window_size, None), offset=0
```
Comment (qdavid1) [inline axlearn/common/flash_attention/tpu_attention_test.py:57]: According to the documentation of splash_attention_mask.LocalMask(), setting window_size=(sliding_window_size, None) would make the mask unbounded on the right, which would not be the same behavior as sliding_window_causal_mask().

Comment (changlan) [reply axlearn/common/flash_attention/tpu_attention_test.py:57]: It is fixed in https://github.com/apple/axlearn/pull/758

Comment (qdavid1) [reply axlearn/common/flash_attention/tpu_attention_test.py:57]: Thanks for fixing!

File: axlearn/common/flash_attention/layer.py [hunk 3]
```diff
@@ -190,11 +205,14 @@ def _compute_attention(
                 )
             attention_logit_biases = attention_logit_biases.astype(q_proj.dtype)
 
-        mask_fn = self._mask_fn
+        if attention_logit_biases is None or self._mask_fn is causal_mask:
```
Comment (berlino) [inline axlearn/common/flash_attention/layer.py:208]: @changlan  I guess this condition should be 

if attention_logit_bias is None or not self._is_mask_fn_used()

Comment (changlan) [reply axlearn/common/flash_attention/layer.py:208]: With today's `_logit_biases_for_mask` impl, it should be safe to just `if attention_logit_bias is None`.
